### ISSUE TITLE
Not Rendering D3.js

### ISSUE THREAD

**Issue by RyanBlundell** [wkhtmltopdf/wkhtmltopdf#4240]
**wkhtmltopdf version(s) affected**: x.y.z 
wkhtmltox-0.12.4_linux-generic-amd64.tar.xz

I have been through a number of threads from people who have had issues rendering D3.js using wkhtmltoimage

I have followed most of the advice:

Ensured all JS/CSS has the full path.

Added javascript delays of as high as 20000ms

Added the following code to the top of the page 

	<script>
	Function.prototype.bind = Function.prototype.bind || function (thisp) {
    var fn = this;
    return function () {
        return fn.apply(thisp, arguments);
    };
};</script>

I have made sure that the div has a specified width / height
I have tried downgrading from D3 version 5 to version 4.

However none of these seem to work. Is there anything else that I can try?

**Comment by RyanBlundell** [comment]
Here is an example of one that is not rendering https://bl.ocks.org/gordlea/raw/27370d1eea8464b04538e6d8ced39e89/

**Comment by zyura** [comment]
You have several critical issues here.

The first one is that your documents are not valid, please don't be lazy to use HTML 5 validator to check your documents: [](https://validator.w3.org/nu/?doc=https%3A%2F%2Fbl.ocks.org%2Fgordlea%2Fraw%2F27370d1eea8464b04538e6d8ced39e89%2F)

The most important is how you try to calculate window size:
```
  , width = window.innerWidth - margin.left - margin.right // Use the window's width 
  , height = window.innerHeight - margin.top - margin.bottom; // Use the window's height
```
This evaluates to "expected" values in modern browsers, but to -100 in wkhtmltopdf.

Just try adding this just after the width/height calculation.
```
document.write('width: ' + width + ', height: ' + height);
```
Basically with wkhtmltopdf you can't rely on innerHeight (or any other height) unless you have it fixed or you have some content within a block.

**Comment by zyura** [comment]
You can't use new fancy JS features with wkhtmltopdf because it's based on the heavily outdated browser engine.

Make sure you replace all arrow functions ```(...) => ...``` with traditional ones ```function() { return ... }```, all ```let``` and ```const``` with ```var```.

Also make sure you get no ```QSslSocket: cannot resolve``` errors in the output (because d3...js) seem to use https to load modules.

You may also need to downgrade d3.v5 to d3.v4: ```http://d3js.org/d3.v4.min.js```.

After making all these changes, it starts producing the correct output.
[d3.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2819219/d3.pdf)

This is the working version of your page: [d3_to_pdf.html](https://gist.github.com/zyura/b6dc1d656bff06b710d187ae4162bd83)

The command line and the output I get (assuming the page is hosted at localhost):
```
wkhtmltopdf --javascript-delay 1000 -s Letter -B 6mm -L 6mm -R 6mm -T 6mm http://localhost/d3_to_pdf.html out.pdf
Loading pages (1/6)
Counting pages (2/6)                                               
Resolving links (4/6)                                                       
Loading headers and footers (5/6)                                           
Printing pages (6/6)
Done
```

**Comment by hvlucas** [comment]
Downgrading jquery and foundation worked for me.
`"foundation-sites": "^6.5.3"` => `"foundation-sites": "^6.4.2"`
`"jquery": "^3.4.1"` => `"jquery": "^3.2.1"`
